pkg/certs/ensure.go

```diff
@@ -96,26 +112,108 @@ func EnsureCerts(
 	return downloadCertsFromSecret(secret, certificateDir)
 }
 
+// downloadCertsFromSecret writes to the filesystem the content of each field in the secret
+// if the field has an equivalent inside the certmap, we write with the corresponding name
+// otherwise the file has the same name than the field
 func downloadCertsFromSecret(
 	secret *corev1.Secret,
 	certificateDir string,
 ) error {
-	for toName, fromName := range certMap {
-		if len(secret.Data[fromName]) == 0 {
-			return fmt.Errorf("secret is missing %s", fromName)
+	certMapValues := maps.Values(certMap)
+	for secretEntry, fileBytes := range secret.Data {
+		name := secretEntry
+		if slices.Contains(certMapValues, secretEntry) {
+			// we need to replace with the actual name
+			for key, sEntry := range certMap {
+				// guarranteed to evaluate to true at least once because of slices.contains
+				if sEntry == secretEntry {
+					if len(fileBytes) == 0 {
+						return fmt.Errorf("secret is missing %s", secretEntry)
+					}
+					name = key
+					break
+				}
+			}
 		}
 
-		name := filepath.Join(certificateDir, toName)
+		name = filepath.Join(certificateDir, name)
 		err := os.MkdirAll(filepath.Dir(name), 0777)
 		if err != nil {
 			return fmt.Errorf("create directory %s", filepath.Dir(name))
 		}
 
-		err = os.WriteFile(name, secret.Data[fromName], 0666)
+		err = os.WriteFile(name, fileBytes, 0666)
 		if err != nil {
-			return fmt.Errorf("write %s: %w", fromName, err)
+			return fmt.Errorf("write %s: %w", name, err)
 		}
 	}
 
 	return nil
 }
+
+func extraFiles(
+	certificateDir string,
+) (map[string][]byte, error) {
+	files := make(map[string][]byte)
+	entries, err := os.ReadDir(certificateDir)
+	if err != nil {
+		return nil, err
+	}
+	for _, v := range entries {
+		if v.IsDir() {
+			// ignore subdirectories for now
+			// etcd files should be picked up by the map
+			continue
+		}
+		name := v.Name()
+
+		// if it's not in the cert map, add to the map
+		_, ok := certMap[name]
+		if !ok {
+			b, err := os.ReadFile(filepath.Join(certificateDir, name))
+			if err != nil {
+				return nil, err
+			}
+			files[name] = b
+		}
+	}
+	return files, err
+}
+
+func WaitForK0sFiles(ctx context.Context, certDir string) (map[string][]byte, error) {
```

Comment:
Why is this function part of this package?

Reply:
because we are waiting for the certs to be created?